File: 2019-kakao-blind-recruitment/Q2.오픈채팅방.py
def solution(record):
    nickname_dic = {}
    query_list = []

    def get_enter_msg(nickname):
        return f"{nickname}님이 들어왔습니다."

    def get_out_msg(nickname):
        return f"{nickname}님이 나갔습니다."

    def handle_query(opt, u_id):
        nickname = nickname_dic[u_id]

        if opt == "Enter":
            return get_enter_msg(nickname)
        else:
            return get_out_msg(nickname)

    for r in record:
        opt, u_id, *rest = r.split(" ")

        if opt in ["Enter", "Change"]:
            nickname = rest[0]
            nickname_dic[u_id] = nickname

        if opt in ["Enter", "Leave"]:
            query_list.append([opt, u_id])

    return [handle_query(*query) for query in query_list]

File: 2019-kakao-blind-recruitment/Q2.오픈채팅방.js
function solution(record) {
	const nicknameMap = new Map();
	const queryList = [];

	const handleSetRecord = ([opt, uID, ...rest]) => {
		if (opt === 'Enter' || opt === 'Change') {
			nicknameMap.set(uID, rest[0]);
		}

		if (opt === 'Enter' || opt === 'Leave') {
			queryList.push([uID, opt]);
		}
	};

	const handleQuery = ([uID, opt]) => {
		const nickName = nicknameMap.get(uID);

		if (opt === 'Enter') {
			return `${nickName}님이 들어왔습니다.`;
		}

		return `${nickName}님이 나갔습니다.`;
	};

	record.map(el => el.split(' ')).forEach(handleSetRecord);
	return queryList.map(handleQuery);
}
